README.md
[![Build Status](https://travis-ci.org/thehelp/client-project.svg?branch=master)](https://travis-ci.org/thehelp/client-project)

# thehelp-client-project

Developing client-side javascript with [`requirejs`](http://requirejs.org/) and testing with [`mocha`](http://visionmedia.github.io/mocha/)? This easy-to-set-up library will streamline your project automation.

## Features

* [`connect`](https://github.com/gruntjs/grunt-contrib-connect) task to serve static files on localhost instead of `file:/` URLs, which have different security settings
* [`grunt-mocha`](https://github.com/kmiyashiro/grunt-mocha) to run your _client-side_ `mocha` tests on the command-line via [`phantomjs`](http://phantomjs.org/)
* [`grunt-requirejs`](https://github.com/asciidisco/grunt-requirejs) to concatenate and optimize your AMD-style modules into production javascript files.
* [`grunt-saucelabs`](https://github.com/axemclion/grunt-saucelabs) to run tests on multiple browsers via [Sauce Labs](https://saucelabs.com)
* 'preamble-for-dist' task to add version number, author, and license information into final distribution files

## Setup

First, install the project (and its friend, [`thehelp-project`](https://github.com/thehelp/project)) as dev dependencies:

```bash
npm install thehelp-project thehelp-client-project --save-dev
```

If you're new to `thehelp-project`, it would be good to read [its docs](https://github.com/thehelp/project) first. Then make sure you have the grunt command available:

```bash
npm install -g grunt-cli
```

## Your Gruntfile

This is the place to start with your Gruntfile:

```javascript
var GruntConfig = require('thehelp-project').GruntConfig;
require('thehelp-client-project').mixin(GruntConfig);

module.exports = function(grunt) {
  var config = new GruntConfig(grunt);

  config.standardSetup();
  config.standardDefault();
};
```

## Usage

### Optimize

These methods give you two easier methods to generate concatenated, optimized javascript files via `requirejs`:

* `registerOptimize()` - Generates a concatenated, optimized file based on AMD modules.
* `registerOptimizeLibrary()` - Generates both minified and non-minified concatenated javascript files. Best-suited for generating a library for consumption by others.

This will generate a single, optimized file named 'home.js' under 'dist/js' which does not need `requirejs` on the page to run, as well as source maps:

```javascript
config.registerOptimize({
  source: 'src/client/home',
  target: 'dist/js/home_standalone.js',
  standalone: true,
  config: { // these are the options piped directly into requirejs
    baseUrl: './',
    paths: {
      backbone: 'lib/vendor/backbone',
      underscore: 'lib/vendor/underscore'
    }
  }
})
```

Or, you can generate a file which isn't standalone. Anything in the `empty` array will not be included in the final file. When your code runs it will still need those packages, so those components must be provided by other files on the page.

```javascript
config.registerOptimize({
  source: 'src/client/home',
  target: 'dist/js/home.js',
  empty: ['util', 'winston'],
  config: {
    baseUrl: './',
    paths: {
      backbone: 'lib/vendor/backbone',
      underscore: 'lib/vendor/underscore'
    }
  }
})
```

`registerOptimizeLibrary` will generate four separate requirejs sub-tasks...

```javascript
config.registerOptimizeLibrary({
  source: 'src/library',
  target: 'library-name',
  targetPath: 'dist/js',
  standalone: true,
  config: {
    baseUrl: './',
    paths: {
      backbone: 'lib/vendor/backbone',
      underscore: 'lib/vendor/underscore'
    }
  }
})
```

...resulting in four files:

* dist/js/library-name.js
* dist/js/library-name.min.js
* dist/js/standalone/library-name.js
* dist/js/standalone/library-name.min.js

The first two files require you to provide a module loader like [`almond.js`](https://github.com/jrburke/almond) or `requirejs`. The second two include `almond.js` and can therefore be used with nothing else on the page.

For more information on requirejs configuration see the integration test included in this project (under 'test/default'), and the `requirejs` documentation:

* [`requirejs` optimization documentation](http://requirejs.org/docs/optimization.html)
* [`requirejs` complete example configuration](https://github.com/jrburke/r.js/blob/master/build/example.build.js)
* [`grunt-requirejs` documentation](https://github.com/asciidisco/grunt-requirejs)

### Add Preamble

Now that you've generated some concatenated and/or optimized files for distribution, you probably want to make sure that people know where the files came from:

```javascript
config.registerPreambleForDist();
config.registerTask('dist', ['requirejs', 'preamble-for-dist']);
```

Elements from your package.json will be injected into javascript files under your dist subdirectory. It will also look for a file LICENSE.txt in your project's root directory to include. And you can add comments to various subsets of your dist files with `options.comments`. More options and details are available at [the detailed documentation.](http://thehelp.github.io/client-project/src/server/mixin.html)

### Test

Another two methods make it easy to automate your client-side testing.

* `registerConnect()` - Sets up a static file server for testing things on the client (called by `standardSetup()` so most of the time you won't call this directly)
* `registerMocha()` - Uses phantomjs to run mocha tests inside a headless browser.

This is how you might create a 'client-test' task which runs your tests against both your development (using `requirejs`) and fully optimized standalone files:

```javascript
config.registerMocha({
  urls: [
    'http://localhost:3001/test/integration/dev.html',
    'http://localhost:3001/test/integration/dist.html'
  ]
});
grunt.registerTask('client-test', ['connect:test', 'mocha']);
```

The default configuration expects that your tests will start whenever they are ready, and will pipe all console output in the browser to the command line. Take a look at the [`grunt-mocha` documentation](https://github.com/kmiyashiro/grunt-mocha) to get into the details.

_Note: the registered `connect:test` task runs the server on port 3001. The `connect:keepalive` task is useful for manual browser-based debugging, and runs on port 3000. This is so you can run grunt test runs while also keeping a server up for your browser-based testing._

### Test on multiple browsers

To get your tests running on multiple browers via the Sauce Labs service, first set two environment variables with information about your account. `thehelp-project` prefers to put them in 'env.json' in your project's root directory:

```json
{
  "SAUCE_USERNAME": "your_username",
  "SAUCE_ACCESS_KEY": "your_key"
}
```

It's likely that you don't want to deploy your files to a publically-accessible server every new build for Sauce Labs, so they've provided [Sauce Labs Connect](https://docs.saucelabs.com/reference/sauce-connect/). It opens a secure tunnel to your machine, giving their services access to ports on your machine. Yes, they take security seriously - they get into the details on the download page.

Once you've got those two things in place, a lot of defaults have been provided for you. This will set things up to run on a minimum set of browsers with reasonable coverage:

```javascript
config.registerSauce({
  urls: [
    'http://localhost:3001/test/integration/dist.html'
  ]
});
grunt.registerTask('cross-browser', ['connect:test', 'sauce']);
```

Various browser subsets are available via `config.saucePlatforms`: `iOS`, `chrome`,
`internetExplorer`, etc. [(complete list of browser subsets).](http://thehelp.github.io/client-project/src/server/sauce_platforms.html) Just set the `browsers` key of the `options` you pass to `registerSauce()`. Check out the additional configuration available at the...

## Detailed Documentation

Detailed docs be found at this project's GitHub Pages, thanks to `groc`: [http://thehelp.github.io/client-project/src/server/mixin.html](http://thehelp.github.io/client-project/src/server/mixin.html)

## Contributing changes

The comprehensive integration test under 'test/default' will be your friend. `./run.sh` and `./clean.sh` will test most of it, but you'll want to visually inspect the dist/js files to ensure that the 'preamble-for-dist' task is doing what it should. The 'cross-browser' task will verify that Sauce Labs integration is still working - you'll need to put your Sauce Labs account information in an env.json file in that directory.

When you have some changes ready, please submit a pull request with:

* Justification - why is this change worthwhile? Link to issues, use code samples, etc.
* Documentation changes for your code updates. Be sure to check the groc-generated HTML with `grunt doc`
* A description of how you tested the change. Don't forget about the very-useful `npm link` command :0)

I may ask you to use a `git rebase` to ensure that your commits are not interleaved with commits already in the history. And of course, make sure `grunt` completes successfully. :0)

## License

(The MIT License)

Copyright (c) 2014 Scott Nonnenberg &lt;scott@nonnenberg.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
## 0.5.0 (2013-03-21)

* `config.saucePlatforms.cheapCoverage` and `config.saucePlatforms.all` updated per [February 2015 statcounter data](http://gs.statcounter.com/) (select mobile browsers as well as specific browser versions, download the CSV).
* Major version updates: `lodash` (2.x to 3.x), `chalk` (0.5.1 to 1.x)
* Minor version updates: `grunt-contrib-connect`, `grunt-saucelabs`
* Update dev dependencies

## 0.4.0 (2014-09-10)

* `grunt-saucelabs` now included with install, set up with `GruntConfig.registerSauce()`
* New 'preamble-for-dist' multi-task to add version number, author, and license information into dist js files. Set up with `GruntConfig.registerPreambleForDist()`

## 0.3.0 (2014-08-28)

* Breaking: removed `registerCopyFromDist()` and `registerCopyFromBower()`; no real need to copy into local repo. Will do it on a case-by-case basis when a project requires a build step.
* Breaking: `registerOptimize()` parameter names made more friendly `name`->`source`, `outName`->`target`, `basePath`->`targetPath`
* travis-ci support
* Documentation updates
* Update `thehelp-project` dev dependency
* Remove `blanket` dev dependency and configuration; we weren't really using it
* Remove `thehelp-test` (circular) dev dependency, add `sinon` and `chai` as replacements
* Full integration test

## 0.2.0 (2014-06-25)

* `registerOptimizeLibrary()` turns off source maps, since they're really big
* Minor version update: `grunt-contrib-connect`
* Update dev dependencies

## 0.1.0 (2014-06-07)

* Six core functions available:
  * `registerConnect()`
  * `registerMocha()`
  * `registerOptimize()`
  * `registerOptimizeLibrary()`
  * `registerCopyFromBower()`
  * `registerCopyFromDist()`


test/default/run.sh
set -e

npm install
bower install
grunt
grunt dist
grunt client-test


test/default/clean.sh
rm -rf dist
rm -rf bower_components
rm -rf docs


test/default/library.js

define(['src/client/home'], function(home) {
  'use strict';

  return home;
});


src/server/util.js
// # Util

'use strict';

var path = require('path');
var fs = require('fs');

function Util() {}

module.exports = Util;

Util.prototype.getPkg = function getPkg() {
  if (this.pkg) {
    return this.pkg;
  }

  try {
    this.pkg = require(path.join(process.cwd(), './package.json'));
  }
  finally {
    this.pkg = this.pkg || {};
    return this.pkg;
  }
};

Util.prototype.getEnv = function getEnv() {
  if (this.env) {
    return this.env;
  }

  try {
    this.env = require(path.join(process.cwd(), './env.json'));
  }
  finally {
    this.env = this.env || {};
    return this.env;
  }
};

Util.prototype.getLicenseContent = function getLicenseContent() {
  if (this.licenseContent) {
    return this.licenseContent;
  }

  try {
    this.licenseContent = fs.readFileSync(path.join(process.cwd(), './LICENSE.txt'));
  }
  finally {
    this.licenseContent = this.licenseContent || '';
    return this.licenseContent;
  }
};


src/server/sauce_platforms.js
/*
# sauce_platforms

Makes it easier to deal with the platforms offered by [Sauce Labs](http://saucelabs.com)
testing systems.

Loosely based on browsers with reasonable usage:
[http://caniuse.com/usage_table.php](http://caniuse.com/usage_table.php)

Sauce Labs current supported platforms:

+ [https://saucelabs.com/platforms](https://saucelabs.com/platforms)
+ (in JSON form) http://saucelabs.com/rest/v1/info/browsers/webdriver

*/

'use strict';

var platforms = module.exports = {};

platforms.safari = [
  ['Mac 10.10', 'safari', '8'],
  ['Mac 10.9', 'safari', '7'],
  ['Mac 10.8', 'safari', '6'],
  ['Mac 10.6', 'safari', '5']
];

platforms.internetExplorer = [
  ['Windows 2012 R2', 'internet explorer', '11'],
  ['Windows 2012', 'internet explorer', '10'],
  ['Windows 2008', 'internet explorer', '9']
];

platforms.iOS = [
  ['Mac 10.10', 'iphone', '8.1'],
  ['Mac 10.10', 'iphone', '7.1']
];

platforms.android = [
  ['Linux', 'android', '5.0'],
  ['Linux', 'android', '4.4'],
  ['Linux', 'android', '4.3'],
  ['Linux', 'android', '4.1'],
  ['Linux', 'android', '4.0']
];

platforms.firefox = [
  ['Windows 2012 R2', 'firefox', '35'],
  ['Windows 2012 R2', 'firefox', '36'],
  ['Windows 2012 R2', 'firefox', '34']
];

platforms.chrome = [
  ['Windows 2012 R2', 'chrome', '40'],
  ['Windows 2012 R2', 'chrome', '39'],
  ['Windows 2012 R2', 'chrome', '37'],
  ['Windows 2012 R2', 'chrome', '31']
];

// ~85-90% of global web usage per Feb 2015 statcounter.com
platforms.all = []
  .concat(platforms.safari)
  .concat(platforms.internetExplorer)
  .concat(platforms.iOS)
  .concat(platforms.android)
  .concat(platforms.firefox)
  .concat(platforms.chrome);

// ~75% of global usage per Feb 2015 statcounter.com (add mobile, select browser versions)
// only missing IE8, whoch our test infrastucture doesn't like
platforms.cheapCoverage = [
  ['Windows 2012 R2', 'chrome', '40'],
  ['Windows 2012 R2', 'firefox', '35'],
  ['Windows 2012 R2', 'internet explorer', '11'],
  ['Mac 10.10', 'iphone', '8.1'],
  ['Windows 2012 R2', 'chrome', '39'],
  ['Windows 2008', 'internet explorer', '9'],
  ['Windows 2012', 'internet explorer', '10'],
  ['Linux', 'android', '5.0'],
  ['Linux', 'android', '4.4']
];



src/server/preamble_renderer.js
// # PreambleRenderer
// The logic for rendering a complete file premble, split out into bite-sized chunks.

'use strict';

var _ = require('lodash');

var Util = require('./util');

function PreambleRenderer(options) {
  options = options || {};

  this.util = options.util || new Util();
}

module.exports = PreambleRenderer;

PreambleRenderer.prototype.renderComments = function(comments, file) {
  var result = '';

  if (comments) {
    _(comments)
      .keys()
      .forEach(function(key) {
        var r = new RegExp(key);
        if (r.test(file)) {
          result += comments[key] + '\n\n';
        }
      })
      .value();
  }

  return result;
};

PreambleRenderer.prototype.renderContributors = function(contributors) {
  var result = '';

  if (contributors) {
    result += 'Contributors:\n';
    _.forEach(contributors, function(contributor) {
      if (typeof contributor === 'string') {
        result += '  ' + contributor + '\n';
      }
      else {
        result += '  ' + contributor.name + ' <' + contributor.email + '>\n';
      }
    });
  }

  return result;
};

PreambleRenderer.prototype.renderLicense = function(licenseName, licenseContent) {
  var result;

  var content = licenseContent || this.util.getLicenseContent();

  if (content) {
    result = '\n\n' + content + '\n';
  }
  else {
    result = '\nLicense: https://spdx.org/licenses/' + licenseName + '\n';
  }

  return result;
};

PreambleRenderer.prototype.go = function(file, options) {
  var result = '';
  var pkg = options.pkg || this.util.getPkg();

  result +=
    '/*' + '\n' +
    pkg.name + ' v' + pkg.version + '\n' +
    options.date + '\n\n';

  result += pkg.homepage + '\n\n';

  result += this.renderComments(options.comments, file);

  result +=
    'Author: ' + pkg.author + '\n';

  result += this.renderContributors(pkg.contributors);

  result += this.renderLicense(pkg.license, options.licenseContent);

  result += '*/\n\n';

  return result;
};


src/server/mixin.js
/*
# mixin

Adds a set of new client-side development configuration methods to the GruntConfig class
provided by `thehelp-project`.
*/

'use strict';

var path = require('path');

var _ = require('lodash');
var chalk = require('chalk');

var Util = require('./util');
var PreambleRenderer = require('./preamble_renderer');
var saucePlatforms = require('./sauce_platforms');

module.exports = function mixin(GruntConfig) {

  GruntConfig.prototype.util = new Util();
  GruntConfig.prototype.saucePlatforms = saucePlatforms;

  // Before we replace `standardSetup` with our version which calls `registerConnect` we
  // save the previous method in an array.
  var previous =
    GruntConfig.prototype.previousSetup = GruntConfig.prototype.previousSetup || [];

  var parent = GruntConfig.prototype.standardSetup;
  previous.push(parent);

  GruntConfig.prototype.standardSetup = function(options) {
    options = options || {};

    parent.call(this, options);

    this.registerConnect(options.connect);
  };

  /*
  `registerConnect` sets up two targets for the `grunt-contrib-connect`
  task, which runs a basic file server. Two different targets are provided:

  + test: On port 3001, this server will stop as soon as the grunt run stops,
  which means that it is only useful for grunt-based testing.
  + keepalive: A server on 3000 that runs until explicitly stopped - good
  for active development.

  */
  GruntConfig.prototype.registerConnect = function(options) {
    this.loadLocalNpm('grunt-contrib-connect', __dirname);

    this.grunt.config('connect', options || {
      test: {
        options: {
          base: '.',
          port: 3001
        }
      },
      keepalive: {
        options: {
          base: '.',
          port: 3000,
          keepalive: true
        }
      }
    });
  };

  // `registerMocha` pulls in `grunt-mocha` which uses `phantomjs`
  // and a custom bridge to run in-browser tests on the command line. You're
  // just responsible for the collection of `urls` to hit. You might consider using
  // the port 3001 urls available with the `connect` task above.
  GruntConfig.prototype.registerMocha = function(options) {
    options = options || {};

    if (!options.urls) {
      throw new Error('Need to provide array of urls to registerMocha()!');
    }

    options.reporter = options.reporter || 'Spec';

    if (typeof options.run === 'undefined') {
      //default to allowing tests to run when they are ready
      options.run = false;
    }

    if (typeof options.log === 'undefined') {
      //default to piping browser console output back to the command line
      options.log = true;
    }

    this.loadLocalNpm('grunt-mocha', __dirname);

    this.grunt.config('mocha', {
      default: {
        options: options
      }
    });
  };

  /*
  `registerOptimizeLibrary` uses `grunt-requirejs` to produce one concatenated file,
  optimized and including source maps by default.

  _Sadly, `preserveLicenseComments` is incompatible with `generateSourceMaps`. So we've
  opted for source maps._
  */
  GruntConfig.prototype.registerOptimize = function(options) {
    /*jshint maxcomplexity: 12 */

    if (!this.requirejsRegistered) {
      this.loadLocalNpm('grunt-requirejs', __dirname);
      this.requirejsRegistered = true;
    }

    options = options || {};
    options = _.cloneDeep(options);
    var config = options.config;

    if (!config) {
      throw new Error('Need to provide options.config with requirejs options!');
    }

    if (!options.source) {
      throw new Error('Need to provide options.source!');
    }

    if (options.standalone) {
      config.almond = true;
    }

    options.target = options.target || options.source;
    config.name = options.source;
    config.out = config.out || options.target;

    if (!config.out) {
      throw new Error('Need to provide options.target or options.config.out!');
    }

    //turning on source maps unless explictly configured
    if (typeof config.optimize === 'undefined' &&
      typeof config.generateSourceMaps === 'undefined' &&
      typeof config.preserveLicenseComments === 'undefined') {

      config.optimize = 'uglify2';
      config.generateSourceMaps = true;
      config.preserveLicenseComments = false;
    }

    if (options.empty) {
      _.forEach(options.empty, function(module) {
        config.paths[module] = 'empty:';
      });
    }

    var name = config.out || '';
    name = name.replace(/\./g, '_');
    var taskName = 'requirejs.' + name + '.options';
    this.grunt.config(taskName, config);
  };

  /*
  `registerOptimizeLibrary` uses `grunt-requirejs` to produce both optimized
  and unoptimized versions of a given library using the r.js optimizer.
  If specified, `standalone` versions of that library can be produced as well
  (using almond), resulting in four total files:

  + library.js
  + library.min.js
  + standalone/library.js
  + standalone/library.min.js

  */
  GruntConfig.prototype.registerOptimizeLibrary = function(options) {
    options = options || {};
    options = _.cloneDeep(options);
    var config = options.config || {};
    var targetPath = options.targetPath || 'dist/js';
    var target = options.target || options.source;
    var standalone = options.standalone;

    //turn off source maps by default, since we're providing unminified files
    if (typeof config.generateSourceMaps === 'undefined') {
      config.generateSourceMaps = false;
      config.preserveLicenseComments = true;
    }

    options.standalone = false;

    //not minified, needs requirejs
    config.optimize = 'none';
    config.out = path.join(targetPath, target + '.js');
    this.registerOptimize(options);

    //minified, needs requirejs
    delete config.optimize;
    config.out = path.join(targetPath, target + '.min.js');
    this.registerOptimize(options);

    if (standalone) {
      options.standalone = true;

      //not minified, standalone with almond.js
      config.optimize = 'none';
      config.out = path.join(targetPath, 'standalone', target + '.js');
      this.registerOptimize(options);

      //minified, standalone with almond.js
      delete config.optimize;
      config.out = path.join(targetPath, 'standalone', target + '.min.js');
      this.registerOptimize(options);
    }
  };

  /*
  `registerSauce` pulls in the `grunt-saucelabs` package and sets up its subtask,
  `saucelabs-mocha` with a number of defaults to make it easy. It pulls values from
  'env.json' and 'package.json' in the current working directory, so if you run the
  `grunt` command from another directory than the root of your project, you'll get
  different behavior.

  _Note: by default this plugin uses
  [Sauce Connect](https://docs.saucelabs.com/reference/sauce-connect/) to make local URLs
  available to the Sauce Labs infrastructure. Set `tunneled = false` to disable that
  behavior._

  ```
  config.registerSauce({
    // required
    urls: [
      'can be local URLs, since sauce connect is used'
    ],

    // optional; defaults are:
    username: 'SAUCE_USERNAME from env.json',
    key: 'SAUCE_ACCESS_KEY from env.json',
    testname: 'name from package.json',
    build: 'version from package.json',
    browsers: [
      'a small subset of supported sauce labs platform that gives good coverage'
    ],
    sauceConfig: {
      // test results are available via their unique url to anyone who has it
      public: 'share'
    },
    pollInterval: 750,
    throttling: 2
  });
  ```

  Check [./sauce_platforms.html](sauce_platforms.js) for more information on the sauce
  platform subsets available. [More information on options you can pass to
  the `grunt-saucelabs` tasks](https://github.com/axemclion/grunt-saucelabs#usage)

  */
  GruntConfig.prototype.registerSauce = function(options) {
    /*jshint maxcomplexity: 13 */

    this.loadLocalNpm('grunt-saucelabs', __dirname);

    options = options || {};

    var env = this.util.getEnv();
    options.username = options.username || env.SAUCE_USERNAME;
    options.key = options.key || env.SAUCE_ACCESS_KEY;

    var pkg = options.pkg || this.util.getPkg();
    options.testname = options.testname || pkg.name;
    options.build = options.build || pkg.version;

    options.browsers = options.browsers || this.saucePlatforms.cheapCoverage;

    if (!options.urls) {
      throw new Error('Need to provide the urls to test!');
    }

    options.sauceConfig = options.sauceConfig || {};
    options.sauceConfig.public = options.sauceConfig.public || 'share';

    options.pollInterval = options.pollInternval || 750;
    options.throttled = options.throttled || 2;

    this.grunt.config('saucelabs-mocha', {
      default: {
        options: options
      }
    });

    this.grunt.registerTask('sauce', ['saucelabs-mocha']);
  };

  /*
  `registerPreambleForDist` installs a task called 'preamble-for-dist' which injects
  project version/license/author information into the top of files in the dist folder. By
  default, it applies to all javascript files under your project 'dist' directory.

  It will attempt to load your project's 'package.json' by loading the file by that name
  at `process.cwd()`. You can override this lookup by providing your own `pkg` object on
  `options`. Otherwise, be careful where you run the `grunt` command.

  ```
  grunt.registerPreambleForDist({
    // this key will override package.json from disk
    pkg: {
      // these keys are required, either from pkg or from your package.json
      name: 'project-name',
      version: '1.4',
      homepage: 'http://my.homepage',
      author: 'Someone',

      // optional; both contributor formats supported
      contributors: [
        'Person 1 <email>',
        {name: 'Person', email: 'email'}
      ],

      // only required if no LICENSE.txt is found and licenseContent not provided
      license: 'a short name, like MIT'
    },

    // optional; the set of files to process
    src: ['dist/js/public/*.js', 'dist/library.css'],

    // optional; overrides LICENSE.text from working directory
    licenseContent: 'complete license text',

    // optional
    comments: {
      library1: 'Library 1 uses components from X, Y and X packages.',
      '*': 'This project is a labor of love'
    }
  });
  ```

  _Note: keys in the `comments` object are regular expressions. If a file matches more
  than one key, each message will be included._

  */
  GruntConfig.prototype.registerPreambleForDist = function(options) {
    var grunt = this.grunt;
    var renderer = new PreambleRenderer({
      util: this.util
    });

    grunt.registerMultiTask('preamble-for-dist', function() {
      var files = this.filesSrc;
      var options = this.options();

      options = options || {};
      options.date = options.date || new Date();

      _.forEach(files, function(file) {
        var contents = grunt.file.read(file);

        contents = renderer.go(file, options) + contents;

        grunt.file.write(file, contents);

        grunt.log.writeln('File ' + chalk.cyan(file) + ' updated.');
      });
    });

    options = options || {};
    options.src = options.src || ['dist/**/*.js'];

    this.grunt.config('preamble-for-dist', {
      default: {
        src: options.src,
        options: options
      }
    });
  };
};
